src/values.js

```diff
@@ -115,7 +115,10 @@ Ref.prototype.inspect = function() {
  * @returns {boolean}
  */
 Ref.prototype.equals = function(other) {
-  return other instanceof Ref && this.value === other.value;
+  return (other instanceof Ref) &&
+    this.id == other.id &&
+    (this.class === undefined || this.class.equals(other.class)) &&
```

Comment:
What if `other` nas a `class` or `database` field and `this` doesn't. Wouldn't that make them different?
The `id` field is probably "equal enough" but, technically, it's best to assert the entire data structure, right?

Reply:
The `equals` check should fail if `other` doesn't have a class or database.

We were asserting the entire structure before, but I don't think it was using the defined `equals` method. As such, Refs with the same structure were failing the check.

Reply:
It think what happened is that the `value` field was the only internal representation of `Ref` when the `equals` method was created. At the time, the `value` field was a string containing the reference id. 

While adding support for recursive references, the `value`'s field type was changed but, the implementation of `equals` wasn't.

My previous comment was more about the fact that on the line 120, if `this.class` is `undefined`, the condition is true regardless if the `other.class` is not `undefined`, which in that case would mean that both references are not equal.

The same goes for database check at line 121.

Does that make sense?

Reply:
Oh, derp. You're right. I see what you're saying now.